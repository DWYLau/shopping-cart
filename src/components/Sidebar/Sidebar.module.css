.flex-container {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  padding: 1rem;
  margin-left: 1rem;
  margin-top: 3rem;
  height: 30vh;
  width: 15vw;
}

.flex-container h2 {
  font-family: system-ui, sans-serif;
  font-size: 2rem;
  font-weight: bold;
  line-height: 1.1;
}

.label {
  font-family: system-ui, sans-serif;
  font-size: 1.4rem;
  font-weight: bold;
  line-height: 1.1;
  display: grid;
  grid-template-columns: 1em auto;
  gap: 0.5em;
}

.label input[type="checkbox"] {
  -webkit-appearance: none;
  appearance: none;
  background-color: #fff;
  margin: 0;
  font: inherit;
  color: black;
  width: 1.15em;
  height: 1.15em;
  border: 0.15em solid black;
  border-radius: 0.15em;
  transform: translateY(-0.075em);
  display: grid;
  place-content: center;
}

.button {
  font-family: system-ui, sans-serif;
  font-size: 1.2rem;
  font-weight: bold;
  background-color: #ccff00;
  height: 10vh;
  width: 40%;
  padding: 0.5rem;
  border-radius: 5px;
}

.button:hover {
  cursor: pointer;
  border: 2px solid #5f9ea0;
}

input[type="checkbox"]::before {
  content: "";
  width: 0.65em;
  height: 0.65em;
  transform: scale(0);
  transition: 120ms transform ease-in-out;
  box-shadow: inset 1em 1em blue;
}

input[type="checkbox"]:checked::before {
  transform: scale(1);
}
