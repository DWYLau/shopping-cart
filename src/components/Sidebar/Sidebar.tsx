import { useState } from "react"
import styles from "./Sidebar.module.css"
import { useSearch } from "../context/SearchContext"

interface Props {
  getCategory: (value: string) => void
}

function Sidebar({ getCategory }: Props) {
  const [category, setCategory] = useState("")
  const { setSearchValue } = useSearch()

  function handleCheck(event: React.ChangeEvent<HTMLInputElement>): void {
    setCategory(event.target.value.toLowerCase())
    getCategory(event.target.value.toLowerCase())
    setSearchValue("")
  }

  function handleReset() {
    setCategory("")
    getCategory("")
    setSearchValue("")
  }

  return (
    <div className={styles["flex-container"]}>
      <h2>Category</h2>
      <label className={styles.label}>
        <input
          type='checkbox'
          name='mens-clothes'
          value="Men's Clothing"
          onChange={handleCheck}
          checked={category === "men's clothing"}
        />
        Men's Clothing
      </label>
      <label className={styles.label}>
        <input
          type='checkbox'
          name='women-clothes'
          value="Women's Clothing"
          onChange={handleCheck}
          checked={category === "women's clothing"}
        />
        Women's Clothing
      </label>
      <label className={styles.label}>
        <input
          type='checkbox'
          name='jewelery'
          value='Jewelery'
          onChange={handleCheck}
          checked={category === "jewelery"}
        />
        Jewelery
      </label>
      <label className={styles.label}>
        <input
          type='checkbox'
          name='electronics'
          value='Electronics'
          onChange={handleCheck}
          checked={category === "electronics"}
        />
        Electronics
      </label>
      <button onClick={handleReset}>Reset</button>
    </div>
  )
}

export default Sidebar
