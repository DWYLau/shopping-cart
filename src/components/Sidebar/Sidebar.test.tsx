import { getByLabelText, render, screen } from "@testing-library/react"
import Sidebar from "./Sidebar"
import { describe, it, expect } from "vitest"
import { BrowserRouter } from "react-router-dom"
import { vi } from "vitest"
import userEvent from "@testing-library/user-event"

describe("Sidebar component", () => {
  const mockGetCategory = vi.fn()

  it("renders correctly", () => {
    const { container } = render(
      <BrowserRouter>
        <Sidebar getCategory={mockGetCategory} />
      </BrowserRouter>
    )
    expect(mockGetCategory).toHaveBeenCalledTimes(0)

    expect(container).toMatchSnapshot()
  })

  it("renders men's clothing label", () => {
    const { container } = render(
      <BrowserRouter>
        <Sidebar getCategory={mockGetCategory} />
      </BrowserRouter>
    )
    expect(mockGetCategory).toHaveBeenCalledTimes(0)

    getByLabelText(container, "Men's Clothing")
  })

  it("checkbox gets clicked and returns true", async () => {
    const user = userEvent.setup()

    render(
      <BrowserRouter>
        <Sidebar getCategory={mockGetCategory} />
      </BrowserRouter>
    )

    const checkbox = screen.getByRole("checkbox", {
      name: "Men's Clothing",
    }) as HTMLInputElement

    await user.click(checkbox)

    expect(checkbox.checked).toEqual(true)
  })

  it("should call getCategory if checkbox is clicked", async () => {
    const mockGetCategory = vi.fn()
    const user = userEvent.setup()

    render(
      <BrowserRouter>
        <Sidebar getCategory={mockGetCategory} />
      </BrowserRouter>
    )

    const checkbox = screen.getByRole("checkbox", {
      name: "Men's Clothing",
    }) as HTMLInputElement

    await user.click(checkbox)

    expect(mockGetCategory).toHaveBeenCalled()
  })

  it("should not call getCategory if checkbox is not clicked", async () => {
    const mockGetCategory = vi.fn()

    render(
      <BrowserRouter>
        <Sidebar getCategory={mockGetCategory} />
      </BrowserRouter>
    )

    expect(mockGetCategory).not.toHaveBeenCalled()
  })
})
