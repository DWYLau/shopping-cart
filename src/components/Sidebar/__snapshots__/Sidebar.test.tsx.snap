// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar component > renders correctly 1`] = `
<div>
  <div
    class="_flex-container_0bec62"
  >
    <h2>
      Category
    </h2>
    <label
      class="_label_0bec62"
    >
      <input
        name="mens-clothes"
        type="checkbox"
        value="Men's Clothing"
      />
      Men's Clothing
    </label>
    <label
      class="_label_0bec62"
    >
      <input
        name="women-clothes"
        type="checkbox"
        value="Women's Clothing"
      />
      Women's Clothing
    </label>
    <label
      class="_label_0bec62"
    >
      <input
        name="jewelery"
        type="checkbox"
        value="Jewelery"
      />
      Jewelery
    </label>
    <label
      class="_label_0bec62"
    >
      <input
        name="electronics"
        type="checkbox"
        value="Electronics"
      />
      Electronics
    </label>
    <button>
      Reset
    </button>
  </div>
</div>
`;
